storage/src/key_value/agent.rs

```diff
@@ -0,0 +1,603 @@
+            &transaction.memorandum[..],
+            &(*memo_index).to_le_bytes()[..],
+        )?;
+        *memo_index += 1;
+
+        Ok(commitments)
+    }
+
+    fn commit_block(&mut self, block_hash: &Digest, ledger_digest: &Digest) -> Result<BlockStatus> {
+        let canon_height = self.canon_height()?;
+        let block = self.get_block(block_hash)?;
+        match self.get_block_state(block_hash)? {
+            BlockStatus::Committed(_) => return Err(StorageError::ExistingCanonBlock(hex::encode(block_hash)).into()),
+            BlockStatus::Unknown => return Err(anyhow!("attempted to commit unknown block")),
+            _ => (),
+        }
+
+        let mut sn_index = self.read_meta_u32(KEY_CURR_SN_INDEX, Some(0))?;
+        let mut cm_index = self.read_meta_u32(KEY_CURR_CM_INDEX, Some(0))?;
+        let mut memo_index = self.read_meta_u32(KEY_CURR_MEMO_INDEX, Some(0))?;
+
+        let mut transaction_cms = vec![];
+
+        for (index, transaction) in block.transactions.iter().enumerate() {
+            let cms = self.commit_transaction(&mut sn_index, &mut cm_index, &mut memo_index, transaction)?;
+            transaction_cms.extend(cms);
+
+            let transaction_location = TransactionLocation {
+                index: index as u32,
+                block_hash: block_hash.clone(),
+            };
+            let mut out = vec![];
+            transaction_location.write(&mut out)?;
+            self.inner
+                .store(KeyValueColumn::TransactionLookup, &transaction.id[..], &out[..])?;
+        }
+
+        self.write_meta_u32(KEY_CURR_SN_INDEX, sn_index)?;
+        self.write_meta_u32(KEY_CURR_CM_INDEX, cm_index)?;
+        self.write_meta_u32(KEY_CURR_MEMO_INDEX, memo_index)?;
+
+        let is_genesis = block.header.previous_block_hash.iter().all(|x| *x == 0) && canon_height == 0;
+
+        let new_best_block_number = if is_genesis { 0 } else { canon_height + 1 };
+
+        self.write_meta_u32(KEY_BEST_BLOCK_NUMBER, new_best_block_number)?;
+
+        let block_num_serialized = &new_best_block_number.to_le_bytes()[..];
+
+        self.inner
+            .store(KeyValueColumn::BlockIndex, &block_hash[..], block_num_serialized)?;
+        self.inner
+            .store(KeyValueColumn::BlockIndex, block_num_serialized, &block_hash[..])?;
+
+        self.inner
+            .store(KeyValueColumn::DigestIndex, &ledger_digest[..], block_num_serialized)?;
+        self.inner
+            .store(KeyValueColumn::DigestIndex, block_num_serialized, &ledger_digest[..])?;
+
+        Ok(BlockStatus::Committed(new_best_block_number as usize))
+    }
+
+    fn decommit_transaction(
+        &mut self,
+        sn_index: &mut u32,
+        cm_index: &mut u32,
+        memo_index: &mut u32,
+        transaction: &SerialTransaction,
+    ) -> Result<()> {
+        for serial in transaction.old_serial_numbers.iter() {
+            self.inner.delete(KeyValueColumn::SerialNumber, &serial[..])?;
+            *sn_index -= 1;
```

Comment:
We need to be careful so that the decommits are always done one-by-one starting from the last block (and we're changing the decommit logic with this PR), otherwise the indices will get borked.

Reply:
that's why we do checked removal from ledger